pyintegration/tests/test_dtypes.py

```diff
@@ -0,0 +1,194 @@
+
+        with self.subTest("int array from Python list"):
+            expected = [1, 2, 3]
+            j_array = dtypes.int32.array_of([1.1, 2.2, 3.3])
+            self.assertIn("[I", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertEqual(expected, py_array)
+
+        with self.subTest("short array from Python list with null_to_value"):
+            py_list = [NULL_SHORT, 1000, 2000, 3000]
+            expected = [0, 1000, 2000, 3000]
+            j_array = dtypes.short.array_of(py_list, null_to_value=0)
+            self.assertIn("[S", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertEqual(expected, py_array)
+
+        with self.subTest("byte array from Python list, down cast"):
+            expected = [1000, 2000, 3000]
+            j_array = dtypes.byte.array_of(expected)
+            self.assertIn("[B", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertNotEqual(expected, py_array)
+
+    def test_floating_array_of(self):
+        with self.subTest("double array from numpy array with nan2value null2value"):
+            np_array = np.array([float('nan'), NULL_DOUBLE, 1.1, float('inf')], dtype=np.float64)
+            expected = [NULL_DOUBLE, 0, 1.1, float('inf')]
+            j_array = dtypes.float64.array_of(np_array, nan_to_value=NULL_DOUBLE, null_to_value=0.0)
+            self.assertIn("[D", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertEqual(expected, py_array)
+
+        with self.subTest("double array from numpy array"):
+            np_array = np.array([float('nan'), NULL_DOUBLE, 1.1, float('inf')], dtype=np.float64)
+            pd_series = pd.Series(np_array)
+            j_array = dtypes.double.array_of(pd_series)
+            py_array = [x for x in j_array]
+            expected = [float('nan'), NULL_DOUBLE, 1.1, float('inf')]
+            self.assertTrue(math.isnan(py_array[0]))
+            self.assertEqual(expected[1:], py_array[1:])
+
+        with self.subTest("float array from numpy array, down cast from double to float"):
+            np_array = np.array([float('nan'), 1.1, float('inf')], dtype=np.float64)
+            expected = [NULL_FLOAT, 1.1, float('inf')]
+            j_array = dtypes.float32.array_of(np_array, nan_to_value=NULL_FLOAT)
+            self.assertIn("[F", str(type(j_array)))
+            py_array = [x for x in j_array]
+            for i in range(3):
+                self.assertAlmostEqual(expected[i], py_array[i])
+
+        with self.subTest("double array from numpy long array"):
+            expected = [NULL_LONG, 1, 2, 3]
+            np_array = np.array(expected, dtype=np.int64)
+            j_array = dtypes.float64.array_of(np_array)
+            self.assertIn("[D", str(type(j_array)))
+            py_array = [x for x in j_array]
+            for i in range(4):
+                self.assertAlmostEqual(expected[i], py_array[i])
```

Comment:
This case illustrates the weirdness in null handling.  Null long values are input, but the numerical float equivalent comes out the other side.